tests/test_state/test_storage_state.py

```diff
@@ -158,3 +301,294 @@ def test_clamp_energy_asserts_battery_traded_more_than_energy_per_slot(self, is_
+        storage_state, current_time_slot = self._setup_registration_test()
+        energy = 1
+        storage_state.offered_buy_kWh[current_time_slot] = 2 * energy
+        storage_state.register_energy_from_bid_trade(
+            energy=energy, time_slot=current_time_slot)
+        storage_state.register_energy_from_bid_trade(
+            energy=energy, time_slot=current_time_slot)
+        assert storage_state.pledged_buy_kWh[current_time_slot] == 2 * energy
+        assert storage_state.offered_buy_kWh[current_time_slot] == 0
+
+    def test_register_energy_from_offer_trade_negative_energy_raise_error(self):
+        storage_state, current_time_slot = self._setup_registration_test()
+        self._assert_negative_energy_raise_error(
+            storage_state.register_energy_from_offer_trade, energy=-1, time_slot=current_time_slot)
+
+    def test_register_energy_from_offer_trade_energy_register_in_dict(self):
+        storage_state, current_time_slot = self._setup_registration_test()
+        energy = 1
+        storage_state.offered_sell_kWh[current_time_slot] = 2 * energy
+        storage_state.register_energy_from_offer_trade(
+            energy=energy, time_slot=current_time_slot)
+        storage_state.register_energy_from_offer_trade(
+            energy=energy, time_slot=current_time_slot)
+        assert storage_state.pledged_sell_kWh[current_time_slot] == 2 * energy
+        assert storage_state.offered_sell_kWh[current_time_slot] == 0
+
+    def _setup_registration_test(self):
+        storage_state = StorageState()
+        current_time_slot, _, _ = self._initialize_time_slots()
+        storage_state.add_default_values_to_state_profiles([current_time_slot])
+        storage_state.activate(
+            slot_length=duration(minutes=15), current_time_slot=current_time_slot)
+        return storage_state, current_time_slot
+
+    @staticmethod
+    def _assert_negative_energy_raise_error(method, energy, time_slot):
+        with pytest.raises(AssertionError):
+            method(energy, time_slot)
+
+    @staticmethod
+    def _setup_storage_state_for_energy_origin_tracking():
+        storage_state = StorageState()
+        used_storage_share = [EnergyOrigin(ESSEnergyOrigin.LOCAL, 0.4),
+                              EnergyOrigin(ESSEnergyOrigin.EXTERNAL, 0.5),
+                              EnergyOrigin(ESSEnergyOrigin.UNKNOWN, 0.1)]
+        storage_state._used_storage_share = used_storage_share
+        return storage_state
+
+    def test_track_energy_bought_type_append_new_energy_origin(self):
+        storage_state = self._setup_storage_state_for_energy_origin_tracking()
+        initial_registry_number = len(storage_state._used_storage_share)
+        storage_state._track_energy_bought_type(energy=1.0, energy_origin=ESSEnergyOrigin.LOCAL)
+        assert len(storage_state._used_storage_share) == initial_registry_number + 1
+        assert isinstance(storage_state._used_storage_share[-1], EnergyOrigin)
```

Comment:
It would be better if you also test for the exact EnergyOrigin enum:
```suggestion
        assert storage_state._used_storage_share[-1].origin == EnergyOrigin. LOCAL
        assert storage_state._used_storage_share[-1].value == energy
```